presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -39,56 +41,81 @@
     private static final double MINIMUM_COMPRESSION_RATIO = 0.8;
 
     private final BlockEncodingSerde blockEncodingSerde;
-    private final Optional<Compressor> compressor;
-    private final Optional<Decompressor> decompressor;
+    @Nullable private final Compressor compressor;
+    @Nullable private final Decompressor decompressor;
+    @Nullable private final SpillCipher spillCipher;
 
-    public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> compressor, Optional<Decompressor> decompressor)
+    public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> compressor, Optional<Decompressor> decompressor, Optional<SpillCipher> spillCipher)
     {
         this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
-        this.compressor = requireNonNull(compressor, "compressor is null");
-        this.decompressor = requireNonNull(decompressor, "decompressor is null");
         checkArgument(compressor.isPresent() == decompressor.isPresent(), "compressor and decompressor must both be present or both be absent");
+        this.compressor = requireNonNull(compressor, "compressor is null").orElse(null);
+        this.decompressor = requireNonNull(decompressor, "decompressor is null").orElse(null);
+        this.spillCipher = requireNonNull(spillCipher, "spillCipher is null").orElse(null);
+        checkState(this.spillCipher == null || !this.spillCipher.isDestroyed(), "spillCipher is already destroyed");
     }
 
     public SerializedPage serialize(Page page)
     {
-        SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact((page.getSizeInBytes() + Integer.BYTES))); // block length is an int
+        SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact(page.getSizeInBytes() + Integer.BYTES)); // block length is an int
         writeRawPage(page, serializationBuffer, blockEncodingSerde);
+        Slice slice = serializationBuffer.slice();
+        int uncompressedSize = serializationBuffer.size();
         byte markings = PageCodecMarker.none();
 
-        if (!compressor.isPresent()) {
-            return new SerializedPage(serializationBuffer.slice(), markings, page.getPositionCount(), serializationBuffer.size());
+        if (compressor != null) {
+            ByteBuffer compressionBuffer = ByteBuffer.allocate(compressor.maxCompressedLength(uncompressedSize));
+            compressor.compress(slice.toByteBuffer(), compressionBuffer);
+            compressionBuffer.flip();
+            if ((((double) compressionBuffer.remaining()) / ((double) uncompressedSize)) <= MINIMUM_COMPRESSION_RATIO) {
```

Comment:
second cast to double is not necessary: `if ((((double) compressionBuffer.remaining()) / uncompressedSize) <= MINIMUM_COMPRESSION_RATIO) {`